assets.go

```diff
@@ -1,354 +1,134 @@
-	if !headless {
-		id = Gl.CreateTexture()
-
-		Gl.BindTexture(Gl.TEXTURE_2D, id)
-
-		Gl.TexParameteri(Gl.TEXTURE_2D, Gl.TEXTURE_WRAP_S, Gl.CLAMP_TO_EDGE)
-		Gl.TexParameteri(Gl.TEXTURE_2D, Gl.TEXTURE_WRAP_T, Gl.CLAMP_TO_EDGE)
-		Gl.TexParameteri(Gl.TEXTURE_2D, Gl.TEXTURE_MIN_FILTER, Gl.LINEAR)
-		Gl.TexParameteri(Gl.TEXTURE_2D, Gl.TEXTURE_MAG_FILTER, Gl.NEAREST)
-
-		if img.Data() == nil {
-			panic("Texture image data is nil.")
+// LoadMany loads the given resources into memory, stopping at the first error.
+func (formats *Formats) LoadMany(urls ...string) error {
+	for _, url := range urls {
+		err := formats.Load(url)
+		if err != nil {
+			return err
 		}
-
-		Gl.TexImage2D(Gl.TEXTURE_2D, 0, Gl.RGBA, Gl.RGBA, Gl.UNSIGNED_BYTE, img.Data())
 	}
-
-	return &Texture{id, float32(img.Width()), float32(img.Height())}
-}
-
-// Width returns the width of the texture.
-func (t *Texture) Width() float32 {
-	return t.width
-}
-
-// Height returns the height of the texture.
-func (t *Texture) Height() float32 {
-	return t.height
-}
-
-func (t *Texture) Texture() *gl.Texture {
-	return t.id
-}
-
-func (r *Texture) View() (float32, float32, float32, float32) {
-	return 0.0, 0.0, 1.0, 1.0
+	return nil
 }
 
-func (r *Texture) Close() {
-	if !headless {
-		Gl.DeleteTexture(r.id)
+// Unload releases the given resource from memory.
+func (formats *Formats) Unload(url string) error {
+	ext := filepath.Ext(url)
+	if loader, ok := Files.formats[ext]; ok {
+		return loader.Unload(url)
 	}
+	return ResourceLoaderNotFoundError{Format: ext, URL: url}
 }
 
-type Sprite struct {
-	Position *Point
-	Scale    *Point
-	Anchor   *Point
-	Rotation float32
-	Color    color.Color
-	Alpha    float32
-	Region   *Region
-}
-
-func NewSprite(region *Region, x, y float32) *Sprite {
-	return &Sprite{
-		Position: &Point{x, y},
-		Scale:    &Point{1, 1},
-		Anchor:   &Point{0, 0},
-		Rotation: 0,
-		Color:    color.White,
-		Alpha:    1,
-		Region:   region,
+// Resource returns the given resource, and an error if it didn't succeed
+func (formats *Formats) Resource(url string) (Resource, error) {
+	ext := filepath.Ext(url)
+	if loader, ok := Files.formats[ext]; ok {
+		return loader.Resource(url)
 	}
+	return nil, ResourceLoaderNotFoundError{Format: ext, URL: url}
 }
 
-func ImageToNRGBA(img image.Image, width, height int) *image.NRGBA {
-	newm := image.NewNRGBA(image.Rect(0, 0, width, height))
-	draw.Draw(newm, newm.Bounds(), img, image.Point{0, 0}, draw.Src)
-
-	return newm
+// A ResourceLoaderNotFoundError is returned whenever the specified file format has no registered `FileLoader` to
+// return the requested resource.
+type ResourceLoaderNotFoundError struct {
+	Format string
+	URL    string
 }
 
-// ImageObject is a pure Go implementation of a `Drawable`
-type ImageObject struct {
-	data *image.NRGBA
+func (r ResourceLoaderNotFoundError) Error() string {
+	return fmt.Sprintf("no `FileLoader` registered for file format %s (within url: %s)", r.Format, r.URL)
```

Comment:
Personal preference, but I'd use `%q` rather than `%s` for the file extension and url, thus resulting in an output along the lines of:

```
no `FileLoader` registered for file format ".png" (within url: "assets/foo.png")
```

Note, for error messages (e.g. `r.Err`), I'd still use `%s`.